### ISSUE TITLE
Cluster State gets stuck with OOM nodes

### ISSUE THREAD

**Issue by pickypg** [elastic/elasticsearch#9018]
Earlier today, we saw a few nodes on a large cluster that hit `OutOfMemoryError`s (but then stayed "alive") while running on 1.3.4. As a result, the `pending_tasks` list filled up and never seemed to seriously progress (unfortunately the _cat variant does not report the "executing" flag, but it should be the first item returned):

```
insertOrder timeInQueue priority  source 
      13670        2.6h IMMEDIATE zen-disco-node_failed([node-3][e183VPoESsScJnZPM7NI0w][node-3][inet[node-3/3.4.5.6:9300]]{ ... master=false ... }), reason failed to ping, tried [3] times, each with maximum [30s] timeout 
      13671        2.6h IMMEDIATE zen-disco-node_failed([node-4][y4UAKUZbQA-Uqc49MwkFCg][node-4][inet[node-3/4.5.6.7:9300]]{ ... master=false ... }), reason failed to ping, tried [3] times, each with maximum [30s] timeout
.
.
.
```

After a long time it seemed to progress, but it's incredibly unclear as to why it took so long. The next run of pending tasks was not for ~30 minutes:

```json
{
  "insert_order" : 13479,
  "priority" : "URGENT",
  "source" : "shard-started ([routing][4], node[Y3Tg53SdTu22zvuT3UOSAg], [R], s[INITIALIZING]), reason [after recovery (replica) from node [[node-1][Gbujn9BhQXSwPrCqZLWlMQ][node-name][inet[/1.2.3.4:9300]]{ ... master=false ... }]]",
  "executing" : true,
  "time_in_queue_millis" : 15988227,
  "time_in_queue" : "4.4h"
}, {
  "insert_order" : 13480,
  "priority" : "URGENT",
  "source" : "shard-started ([rollup-summarized-2014-12-10][9], node[0cayvawQRSSQEpcGAxFRhQ], [R], s[INITIALIZING]), reason [after recovery (replica) from node [[node-2][y4UAKUZbQA-Uqc49MwkFCg][node-2][inet[/2.3.4.5:9300]]{ ... master=false ... }]]",
  "executing" : false,
  "time_in_queue_millis" : 15986821,
  "time_in_queue" : "4.4h"
}
```

I have the complete list upon request (internal 6345 and 6402).

The solution to the problem was to bounce every node that we discovered to have `OutOfMemoryError`s -- there were at least 5. Obviously those nodes were a problem, but getting the rest of the cluster stuck is also a problem.

**Comment by aswan** [comment]
I just ran into something that sounds similar, stack trace below.  This is followed by a variety of warnings in the logs, all of which seem to be variants of "my connection with this other node is broken".  Sorry for the dumb question, I'm unfamiliar with the internals here, is there any hope of making this more resilient?

```
[2015-03-26 00:22:52,695][DEBUG][action.search.type       ] [node-984cdcf3-a8ed-4fd2-87b1-c3965a60e9d3] [events-default@2015.03.19][1], node[IZDRTRixQ823AXpcGm7KNg], [P], s[STARTE
D]: Failed to execute [org.elasticsearch.action.search.SearchRequest@37e55556] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: Failed to deserialize response of type [org.elasticsearch.search.query.QuerySearchResult]
Caused by: org.elasticsearch.transport.TransportSerializationException: Failed to deserialize response of type [org.elasticsearch.search.query.QuerySearchResult]
        at org.elasticsearch.transport.netty.MessageChannelHandler.handleResponse(MessageChannelHandler.java:152)
        at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:127)
        at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:296)
        at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
        at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
        at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:310)
        at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
        at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:268)
        at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:255)
        at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
        at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
        at org.elasticsearch.common.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
        at org.elasticsearch.common.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
[2015-03-26 00:22:52,702][WARN ][transport.netty          ] [node-984cdcf3-a8ed-4fd2-87b1-c3965a60e9d3] exception caught on transport layer [[id: 0xc69b3bd8, /172.31.9.100:46763 => /172.31.10.126:9300]], closing connection
java.lang.OutOfMemoryError: Java heap space
[2015-03-26 00:22:52,696][WARN ][netty.channel.socket.nio.AbstractNioSelector] Unexpected exception in the selector loop.
java.lang.OutOfMemoryError: Java heap space
[2015-03-26 00:22:52,696][WARN ][netty.channel.socket.nio.AbstractNioSelector] Unexpected exception in the selector loop.
java.lang.OutOfMemoryError: Java heap space
        at java.util.HashMap.newKeyIterator(HashMap.java:968)
        at java.util.HashMap$KeySet.iterator(HashMap.java:1002)
        at java.util.HashSet.iterator(HashSet.java:170)
        at sun.nio.ch.Util$2.iterator(Util.java:303)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:102)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
        at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
        at org.elasticsearch.common.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
        at org.elasticsearch.common.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by mjball** [comment]
I'm also running into OOMs leaving nodes in something of a zombie state. Stack traces:

```
[2015-05-06 14:38:35,028][WARN ][action.index             ] [constance.iad02.hubspot-networks.net] Failed to perform indices:data/write/index on remote replica [slushpond.iad02.hubspot-networks.net][OFjx5P7iSE-YLi4AVf76tg][slushpond][inet[/172.16.12.59:9300]]{availability_zone=us-east-1a, aws_availability_zone=us-east-1a, master=true}[company-event-1][4]
org.elasticsearch.transport.RemoteTransportException: Failed to deserialize exception response from stream
Caused by: org.elasticsearch.transport.TransportSerializationException: Failed to deserialize exception response from stream
	at org.elasticsearch.transport.netty.MessageChannelHandler.handlerResponseError(MessageChannelHandler.java:173)
	at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:125)
	at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
	at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318)
	at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.elasticsearch.common.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.elasticsearch.common.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:48:14,151][WARN ][action.index             ] [constance.iad02.hubspot-networks.net] Failed to send response for indices:data/write/index
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:43:09,828][WARN ][transport.netty          ] [constance.iad02.hubspot-networks.net] Failed to send error message back to client for action [indices:data/write/index[r]]
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:42:11,943][WARN ][index.engine.internal    ] [constance.iad02.hubspot-networks.net] [company-event-1][3] failed engine [out of memory]
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:39:02,693][WARN ][transport.netty          ] [constance.iad02.hubspot-networks.net] Failed to send error message back to client for action [indices:data/write/index[r]]
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:48:33,749][WARN ][index.engine.internal    ] [constance.iad02.hubspot-networks.net] [biden-company-3][4] failed engine [indices:data/write/bulk[s] failed on replica]
org.elasticsearch.index.engine.IndexFailedEngineException: [biden-company-3][4] Index failed for [company#swisstrades.ch]
	at org.elasticsearch.index.engine.internal.InternalEngine.index(InternalEngine.java:530)
	at org.elasticsearch.index.shard.service.InternalIndexShard.index(InternalIndexShard.java:425)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnReplica(TransportShardBulkAction.java:590)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$ReplicaOperationTransportHandler.messageReceived(TransportShardReplicationOperationAction.java:246)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$ReplicaOperationTransportHandler.messageReceived(TransportShardReplicationOperationAction.java:225)
	at org.elasticsearch.transport.netty.MessageChannelHandler$RequestHandler.run(MessageChannelHandler.java:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.lucene.store.AlreadyClosedException: this IndexWriter is closed
	at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:698)
	at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:712)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1507)
	at org.elasticsearch.index.engine.internal.InternalEngine.innerIndex(InternalEngine.java:603)
	at org.elasticsearch.index.engine.internal.InternalEngine.index(InternalEngine.java:522)
	... 8 more
Caused by: java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:48:33,745][WARN ][action.index             ] [constance.iad02.hubspot-networks.net] Failed to perform indices:data/write/index on remote replica [slushpond.iad02.hubspot-networks.net][OFjx5P7iSE-YLi4AVf76tg][slushpond][inet[/172.16.12.59:9300]]{availability_zone=us-east-1a, aws_availability_zone=us-east-1a, master=true}[company-event-1][4]
org.elasticsearch.transport.RemoteTransportException: Failed to deserialize exception response from stream
Caused by: org.elasticsearch.transport.TransportSerializationException: Failed to deserialize exception response from stream
	at org.elasticsearch.transport.netty.MessageChannelHandler.handlerResponseError(MessageChannelHandler.java:173)
	at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:125)
	at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
	at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318)
	at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.elasticsearch.common.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.elasticsearch.common.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:50:23,521][WARN ][cluster.action.shard     ] [constance.iad02.hubspot-networks.net] [company-event-1][4] sending failed shard for [company-event-1][4], node[OFjx5P7iSE-YLi4AVf76tg], [R], s[STARTED], indexUUID [yevQ6omxSmiyjZZKjy1lUw], reason [Failed to perform [indices:data/write/index] on replica, message [RemoteTransportException[Failed to deserialize exception response from stream]; nested: TransportSerializationException[Failed to deserialize exception response from stream]; nested: OutOfMemoryError[Java heap space]; ]]
[2015-05-06 14:50:23,521][WARN ][cluster.action.shard     ] [constance.iad02.hubspot-networks.net] [company-event-1][4] received shard failed for [company-event-1][4], node[OFjx5P7iSE-YLi4AVf76tg], [R], s[STARTED], indexUUID [yevQ6omxSmiyjZZKjy1lUw], reason [Failed to perform [indices:data/write/index] on replica, message [RemoteTransportException[Failed to deserialize exception response from stream]; nested: TransportSerializationException[Failed to deserialize exception response from stream]; nested: OutOfMemoryError[Java heap space]; ]]
[2015-05-06 14:48:33,745][DEBUG][action.bulk              ] [constance.iad02.hubspot-networks.net] [biden-company-3][2], node[O-FiNl4vSJ2qtETa3gqrig], [P], s[STARTED]: Failed to execute [org.elasticsearch.action.bulk.BulkShardRequest@776cc9ab]
java.lang.OutOfMemoryError: Java heap space
	at java.lang.AbstractStringBuilder.<init>(AbstractStringBuilder.java:64)
	at java.lang.StringBuilder.<init>(StringBuilder.java:97)
	at org.elasticsearch.common.logging.support.LoggerMessageFormat.format(LoggerMessageFormat.java:52)
	at org.elasticsearch.common.logging.support.AbstractESLogger.debug(AbstractESLogger.java:71)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:220)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$AsyncShardOperationAction.performOnPrimary(TransportShardReplicationOperationAction.java:512)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$AsyncShardOperationAction$1.run(TransportShardReplicationOperationAction.java:419)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
[2015-05-06 14:48:14,617][WARN ][transport.netty          ] [constance.iad02.hubspot-networks.net] Actual Exception
org.elasticsearch.index.engine.VersionConflictEngineException: [company-event-1][3] [twitterevent][595957573871915008]: version conflict, current [3], provided [2]
	at org.elasticsearch.index.engine.internal.InternalEngine.innerIndex(InternalEngine.java:582)
	at org.elasticsearch.index.engine.internal.InternalEngine.index(InternalEngine.java:522)
	at org.elasticsearch.index.shard.service.InternalIndexShard.index(InternalIndexShard.java:425)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnReplica(TransportIndexAction.java:234)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$ReplicaOperationTransportHandler.messageReceived(TransportShardReplicationOperationAction.java:246)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$ReplicaOperationTransportHandler.messageReceived(TransportShardReplicationOperationAction.java:225)
	at org.elasticsearch.transport.netty.MessageChannelHandler$RequestHandler.run(MessageChannelHandler.java:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
[2015-05-06 14:48:14,616][DEBUG][action.admin.indices.stats] [constance.iad02.hubspot-networks.net] [company-event-1][1], node[O-FiNl4vSJ2qtETa3gqrig], [P], s[STARTED]: failed to execute [org.elasticsearch.action.admin.indices.stats.IndicesStatsRequest@40fc59d1]
java.lang.OutOfMemoryError: Java heap space
	at java.util.HashMap.createEntry(HashMap.java:897)
	at java.util.HashMap.addEntry(HashMap.java:884)
	at java.util.HashMap.put(HashMap.java:505)
	at java.util.HashSet.add(HashSet.java:217)
	at org.apache.lucene.store.FileSwitchDirectory.listAll(FileSwitchDirectory.java:88)
	at org.apache.lucene.store.FilterDirectory.listAll(FilterDirectory.java:48)
	at org.elasticsearch.index.store.DistributorDirectory.listAll(DistributorDirectory.java:88)
	at org.apache.lucene.store.FilterDirectory.listAll(FilterDirectory.java:48)
	at org.elasticsearch.common.lucene.Directories.estimateSize(Directories.java:40)
	at org.elasticsearch.index.store.Store.stats(Store.java:216)
	at org.elasticsearch.index.shard.service.InternalIndexShard.storeStats(InternalIndexShard.java:540)
	at org.elasticsearch.action.admin.indices.stats.CommonStats.<init>(CommonStats.java:134)
	at org.elasticsearch.action.admin.indices.stats.ShardStats.<init>(ShardStats.java:49)
	at org.elasticsearch.action.admin.indices.stats.TransportIndicesStatsAction.shardOperation(TransportIndicesStatsAction.java:205)
	at org.elasticsearch.action.admin.indices.stats.TransportIndicesStatsAction.shardOperation(TransportIndicesStatsAction.java:56)
	at org.elasticsearch.action.support.broadcast.TransportBroadcastOperationAction$AsyncBroadcastAction$1.run(TransportBroadcastOperationAction.java:171)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
[2015-05-06 14:48:14,156][WARN ][action.index             ] [constance.iad02.hubspot-networks.net] Failed to send response for indices:data/write/index
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:48:14,153][WARN ][cluster.action.shard     ] [constance.iad02.hubspot-networks.net] [company-event-1][4] sending failed shard for [company-event-1][4], node[OFjx5P7iSE-YLi4AVf76tg], [R], s[STARTED], indexUUID [yevQ6omxSmiyjZZKjy1lUw], reason [Failed to perform [indices:data/write/index] on replica, message [RemoteTransportException[Failed to deserialize exception response from stream]; nested: TransportSerializationException[Failed to deserialize exception response from stream]; nested: OutOfMemoryError[Java heap space]; ]]
[2015-05-06 15:07:25,948][WARN ][cluster.action.shard     ] [constance.iad02.hubspot-networks.net] [company-event-1][4] received shard failed for [company-event-1][4], node[OFjx5P7iSE-YLi4AVf76tg], [R], s[STARTED], indexUUID [yevQ6omxSmiyjZZKjy1lUw], reason [Failed to perform [indices:data/write/index] on replica, message [RemoteTransportException[Failed to deserialize exception response from stream]; nested: TransportSerializationException[Failed to deserialize exception response from stream]; nested: OutOfMemoryError[Java heap space]; ]]
[2015-05-06 15:07:25,948][WARN ][transport.netty          ] [constance.iad02.hubspot-networks.net] Message not fully read (response) for [20311418] handler org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$AsyncShardOperationAction$4@18821df1, error [true], resetting
[2015-05-06 15:07:25,945][WARN ][http.netty               ] [constance.iad02.hubspot-networks.net] Caught exception while handling client http traffic, closing connection [id: 0x7230c876, /63.139.127.4:45729 => /172.16.45.251:9200]
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 15:07:25,815][WARN ][index.engine.internal    ] [constance.iad02.hubspot-networks.net] [biden-company-3][1] failed engine [out of memory]
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 15:07:25,810][WARN ][netty.channel.socket.nio.AbstractNioSelector] Unexpected exception in the selector loop.
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 15:07:25,805][WARN ][netty.channel.socket.nio.AbstractNioSelector] Unexpected exception in the selector loop.
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:59:32,277][WARN ][http.netty               ] [constance.iad02.hubspot-networks.net] Caught exception while handling client http traffic, closing connection [id: 0xa850e05b, /63.139.127.4:28283 => /172.16.45.251:9200]
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:57:48,450][WARN ][http.netty               ] [constance.iad02.hubspot-networks.net] Caught exception while handling client http traffic, closing connection [id: 0x9ea64b72, /63.139.127.4:54841 => /172.16.45.251:9200]
java.lang.OutOfMemoryError: Java heap space
[2015-05-06 14:57:39,919][WARN ][index.search.slowlog.query] [constance.iad02.hubspot-networks.net] [biden-company-3][3] took[29.9m], took_millis[1794652], types[company], stats[], search_type[QUERY_THEN_FETCH], total_shards[5], source[{"from":0,"size":15,"query":{"bool":{"must":{"bool":{"should":{"bool":{"must":[{"terms":{"countrycode":["SE"]}},{"terms":{"industrycode":["95"]}}]}}}}}},"fields":[],"sort":[{"numcontacts":{"order":"desc","missing":"_last"}},{"domain":{"order":"desc","missing":"_last"}}]}], extra_source[], 
[2015-05-06 14:52:15,181][WARN ][transport.netty          ] [constance.iad02.hubspot-networks.net] Message not fully read (response) for [20311347] handler org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$AsyncShardOperationAction$4@38f80dd, error [true], resetting
[2015-05-06 14:51:53,118][WARN ][index.shard.service      ] [constance.iad02.hubspot-networks.net] [biden-company-3][4] Failed to perform scheduled engine refresh
org.elasticsearch.index.engine.RefreshFailedEngineException: [biden-company-3][4] Refresh failed
	at org.elasticsearch.index.engine.internal.InternalEngine.refresh(InternalEngine.java:831)
	at org.elasticsearch.index.shard.service.InternalIndexShard.refresh(InternalIndexShard.java:498)
	at org.elasticsearch.index.shard.service.InternalIndexShard$EngineRefresher$1.run(InternalIndexShard.java:960)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
```

Full logs: https://gist.github.com/mjball/d9dd9ebd44c49306bc1d

I also took a heap dump but it was not particularly revealing. Happy to send it along if it would be useful. All I saw were 4 ~391 mb `QuerySearchResult` instances. Despite setting `-XX:+HeapDumpOnOutOfMemoryError` it looks like a heap dump was not actually created; I'm not sure why.

**Comment by Srinathc** [comment]
Have been hitting up a lot of OOM issues with release 1.7.0.
Discussions and test results available on the discussion thread
https://discuss.elastic.co/t/merging-of-segments-results-in-java-lang-outofmemoryerror-java-heap-space/26301

**Comment by clintongormley** [comment]
The zombie nodes problem originally reported here was fixed back in 1.4 I believe. Closing

**Comment by pickypg** [comment]
(It was with respect to this issue and I have not seen it since. The issue posted just above is unrelated.)
